doc/developer/design/20210831_correctness.md

```diff
@@ -0,0 +1,124 @@
+The storage layer does not need to capture the entire contents of a collection to make it definite; durable "offset"-to-timestamp bindings can suffice if the source data are durable but without timestamps, and no data may be required if the source data are durable with timestamps (e.g. MzCDC sources).
+Update streams that are definite form the "durable storage" of Materialize, and are the basis for Materialize's durability guarantees.
+
+The **compute layer** (`dataflow`, plus potentially others) consumes streams of updates to collections, and computes and then incrementally maintains the results of views over these data.
+A view is ideally "deterministic" in the sense that it acts as a function on the input update streams, and produces a well-defined output on definite input (we can allow "non-deterministic" views but like "indefinite streams" they taint their results).
+One correctness goal of the `dataflow` module is to convert definite inputs into definite outputs, which can be returned to the storage layer or to interactive user sessions.
+Views and their dependencies form the "concurrency control" of Materialize, in that outputs are reported as definite only when, and ideally as soon as, their inputs are definite.
+
+The **coordination layer** (`coord`, plus potentially others) interfaces with user sessions and pilots the storage and compute layers.
+It maintains mappings from user names to storage and compute objects (source, views, indexes), and dereferences them in ways that provide a consistent experience for users.
+It also chooses timestamps for reads (peeks) and writes (table inserts) to provide the lived experience of continually moving forward through time.
+It pilots the storage and compute layers and is responsible for ensuring they communicate clearly and completely with each other.
+One correctness goal of the `coord` module is to provide "strict sequential consistency" to the interactive users, through the careful use of the storage and compute layers.
+Another correctness goal of the `coord` module is to ensure that the storage and compute correctly and completely interoperate (e.g. that the compute layer is invoked only on definite inputs, and fully communicates its definite outputs back to storage if appropriate).
+
+### Approaches
+
+The plan of record is that source data are made definite by the storage layer, and results are definite when observed through the compute layer (applied to definite inputs).
+Definite data are both durable, and ordered in time (strictly sequentially consistent), and reflect the stated "order" the system presents as having undergone.
+
+The coordination layer has several constraints on its behavior to present as strictly sequentially consistent.
+The coordination layer should only report definite data, or indefinite data that are clearly presented as such.
```

Comment:
How would indefinite data be presented as indefinite? Could this mean, for example, that a user could ask in their session about the definiteness of some data? Or that the coordination layer complains about indefinite data unless explicitly told it's ok to report it? Doesn't need to be specified here, just trying to understand what this might mean.

Reply:
I'm not sure, tbh. I could imagine that, like in the VOLATILE design, there is a column that reports this about sources, views, queryable things. Maybe no more is said than that. The coordinator complains loudly if one attempts to create definite data out of indefinite sources other than by handing it to `persist`.

I'm also less clear on what one says about indefinite data other than `indefinite`. We could say the same things as for definite data: `since` and `upper` which .. just have a less clear meaning now: "I'll answer queries in that time interval", or at least a less clear answer to: "why not answer things outside that interval?"

Perhaps these should be orthogonal, where a collection is either VOLATILE or DEFINITE, and independently there is the available time interval, framed by `since` and `upper`. The meaning is ideally clear for DEFINITE, and VOLATILE with those bounds is a bit more about which queries can be asked, but without guarantee about the nature of the results.

I definitely didn't meant to prescribe SQL level changes to commands in the support of "clearly presented as such", so much as "the coordinator shouldn't confuse the user about which is which".